icu4c/source/i18n/dtptngen.cpp

```diff
@@ -809,12 +813,16 @@ DateTimePatternGenerator::getCalendarTypeToUse(const Locale& locale, CharString&
             calendarType,
```

Comment:
I don't know the common practice in ICU. Is uloc_getKeywordValue designed to handle U_MISSING_RESOURCE_ERROR or any general error code properly? 

Looking at the current implementation of uloc_getKeywordValue, I think it would almost return quickly if !U_SUCCESS(*status). Mentioning as I don't know the C / ICU4C code style in general

Reply:
I think the common pattern in ICU is that if the incoming status/error code is already an error (ie: `U_FAILURE(status) == true`) then the function should do almost nothing and quickly return, possibly returning some sort of default or bogus values for any out-parameters.
In the case of `uloc_getKeywordValue`, if the incoming status is already a failure, then it does effectively nothing and  returns back 0.
In the case of the default root locale, it means that the `localeWithCalendarKey` array will still be empty, the function won't add the "`calendar=gregorian`" part to it.
I purposely did not zero/clear the error code *before* calling this method, in order to keep the behavior exactly the same as what the previous code would have done. (It completely ignored the error from `ures_getFunctionalEquivalent`, and then called `uloc_getKeywordValue`, meaning the array wouldn't get the "calendar=gregorian" added).